Fix/deprecate freeform type bad type

roberth opened the pull request:
- An alternative to https://github.com/NixOS/nixpkgs/pull/438558

I believe this provides a clearer fix with fewer workarounds.

We may further simplify this by removing the first commit, which is a very niche corner case that might not even help anyone.

## TODO

- [ ] tests

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform:
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- Tested, as applicable:
  - [ ] [NixOS tests] in [nixos/tests].
  - [ ] [Package tests] at `passthru.tests`.
  - [ ] Tests in [lib/tests] or [pkgs/test] for functions and "core" functionality.
- [ ] Ran `nixpkgs-review` on this PR. See [nixpkgs-review usage].
- [ ] Tested basic functionality of all binary files, usually in `./result/bin/`.
- Nixpkgs Release Notes
  - [ ] Package update: when the change is major or breaking.
- NixOS Release Notes
  - [ ] Module addition: when adding a new NixOS module.
  - [ ] Module update: when the change is significant.
- [ ] Fits [CONTRIBUTING.md], [pkgs/README.md], [maintainers/README.md] and other READMEs.

[NixOS tests]: https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests
[Package tests]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests
[nixpkgs-review usage]: https://github.com/Mic92/nixpkgs-review#usage

[CONTRIBUTING.md]: https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md
[lib/tests]: https://github.com/NixOS/nixpkgs/blob/master/lib/tests
[maintainers/README.md]: https://github.com/NixOS/nixpkgs/blob/master/maintainers/README.md
[nixos/tests]: https://github.com/NixOS/nixpkgs/blob/master/nixos/tests
[pkgs/README.md]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md
[pkgs/test]: https://github.com/NixOS/nixpkgs/blob/master/pkgs/test

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

roberth committed:
lib.types.either: Make impossible value a warning for now

This helps users fix their code.
Found in https://github.com/NixOS/nixpkgs/issues/438459
A nicer fix for the freeformType comes in the next commit.

roberth committed:
lib.modules.evalModules: Use v2 merge of freeformType and fix+deprecate bad types

Allow bad types (not in attrsOf) to be used in freeformType again, but
add a deprecation message.

lib.either is special-cased, because it provides a less useful message
for its own problem, which is best to ignore when that occurs in
a freeformType.

hsjobeki reviewed (dismissed):
I'm not sure if we should print non-actionable error/warning messages

hsjobeki reviewed (commented):
I think i am in favor of this solution rather than #438558. The risk of changing merge further is currently to big, and feels a bit rushed, while this is the correct solution. Use `mergeDefinitions` fixing the merge behavior of freeform.

But also changing the freeform merge has the risk to reveal other regressions. That relied on freeform beeing unchecked.

roberth committed:
lib/modules: Dedup freeformType context

This way we avoid blasting users with a ton of useless context many times in the
same eval.
